src/Microsoft.AspNetCore.Http.Abstractions/Extensions/EndpointModel.cs

```diff
@@ -0,0 +1,35 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    /// <summary>
+    /// A base class for modeling <see cref="Endpoint"/>.
+    /// </summary>
+    public abstract class EndpointModel
+    {
+        /// <summary>
+        /// Gets or sets the delegate used to process requests for the endpoint.
+        /// </summary>
+        public RequestDelegate RequestDelegate { get; set; }
+
+        /// <summary>
+        /// Gets or sets the informational display name of this endpoint.
+        /// </summary>
+        public string DisplayName { get; set; }
+
+        /// <summary>
+        /// Gets the collection of metadata associated with this endpoint.
+        /// </summary>
+        public IList<object> Metadata { get; } = new List<object>();
```

Comment:
Is there are specific reason to only expose the interface (except for making an abstraction of the underlying implementation)? This isn't a standalone case, I know.

Now that devirtualization is implemented in the JIT this may no longer be a problem though.

Reply:
It's standard to expose an interface publicly in case the internal implementation changes.

For cases like this it doesn't make a huge amount of sense, but this also isn't a type that will be used when perf is important 🤷‍♂️